### ISSUE TITLE
Deleting, moving or renaming shared files during partial sharing does not work

### ISSUE THREAD

**Issue by tobous** [saros-project/saros#235]
Saros/E does not correctly propagate the deletion, move or renaming of shared files to other participants during partial sessions. This leaves the session in an inconsistent state, as the other participants still have the shared files that were deleted/moved/renamed locally. Furthermore, these inconsistencies don't seem to be picked up by the recovery mechanism.

The problem seems to be on the host side, as the host always logs the following error, even though the file is shared and other activities on the file (like text edits or saves) are shared correctly. For example for deletions:
```
ERROR 01:31:05,709 [DispatchContext] (SarosSession.java:839) PSFIR - file removal detected for a non shared file: L/alp/src/de/fu_berlin/alp4/Created.java (EclipseFileImpl)
```

The deletion activity seems to be triggered correctly but then fails during dispatching:
```
TRACE 01:31:05,693 [ModalContext] (SharedResourcesManager.java:223) received resource change event caused by job  ='Implicit Job'
TRACE 01:31:05,693 [ModalContext] (SharedResourcesManager.java:234) received resource delta 'ResourceDelta(/)' contains:
C(0) /
C(0) /alp
C(0) /alp/src
C(0) /alp/src/de
C(0) /alp/src/de/fu_berlin
C(0) /alp/src/de/fu_berlin/alp4
R /alp/src/de/fu_berlin/alp4/Created.java

TRACE 01:31:05,694 [ModalContext] (SharedResourcesManager.java:282) generated resource activities for current resource delta 'ResourceDelta(/)' : [FileActivity [dst:path=SPath [project=P/alp (EclipseProjectImpl), projectRelativePath=src/de/fu_berlin/alp4/Created.java], src:path=N/A, type=REMOVED, encoding=N/A, content=0 byte(s)]]
TRACE 01:31:05,708 [main] (ActivitySequencer.java:373) dispatching activity FileActivity [dst:path=SPath [project=P/alp (EclipseProjectImpl), projectRelativePath=src/de/fu_berlin/alp4/Created.java], src:path=N/A, type=REMOVED, encoding=N/A, content=0 byte(s)] to the local user: intellij_test_1@saros-con.imp.fu-berlin.de/Saros
ERROR 01:31:05,709 [DispatchContext] (SarosSession.java:839) PSFIR - file removal detected for a non shared file: L/alp/src/de/fu_berlin/alp4/Created.java (EclipseFileImpl)
```

The same also applies for moving shared files:
```
ERROR 01:50:12,837 [DispatchContext] (SarosSession.java:861) PSFIM - file move detected for a non shared file, source file is not shared, src: L/alp/src/de/fu_berlin/alp4/Main.java (EclipseFileImpl) , dest: L/alp/src/test/Main.java (EclipseFileImpl)
```
and for renaming shared files:
```
ERROR 01:57:52,281 [DispatchContext] (SarosSession.java:861) PSFIM - file move detected for a non shared file, source file is not shared, src: L/alp/src/de/fu_berlin/alp4/Created.java (EclipseFileImpl) , dest: L/alp/src/de/fu_berlin/alp4/Created_2.java (EclipseFileImpl)
```

#### Steps to Reproduce:
1. Alice and Bob start a partial session that shares the file A.
2. Alice deletes (or moves/renames) the file A.

**Comment by srossbach** [comment]
I am pretty sure this is a regression. However I added some checks (the one you are seeing now) and it seems that is the reason it do not work anymore.

But I am still wondering why the deletion will fail. There is no method in the ISarosSession interface to remove / mark resources as not shared so this error should not happen unless the logic was executed before or the isShared method is broken.

**Comment by srossbach** [comment]
Ok it took me awhile to understand how this activity stuff works, especially the interaction between the SarosSession, ActivitySequencer, and ActivityHandler

However here is the fix:


SarosSession.java

@@ -754,9 +754,19 @@
         }
 
         boolean send = true;
-        // handle FileActivities and FolderActivities to update ProjectMapper
+
+        /*
+         * To explain some magic: When the host generates an activity this
+         * method will be called multiple times. It is first called with the
+         * host as it is only recipient and then it is called again for the
+         * other recipients.
+         */
+        // handle IFileSystemModificationActivities to update ProjectMapper
         if (activity instanceof IFileSystemModificationActivity) {
-            send = updatePartialSharedResources((IFileSystemModificationActivity) activity);
+            if (isHost() && recipients.contains(getLocalUser())) {
+                assert recipients.size() == 1 : "There should only be one target if the host fires an original activity";
+                send = updatePartialSharedResources((IFileSystemModificationActivity) activity);
+            }
         }
 
         if (!send)

Would you please so kind and apply it to the master ? @tobous

**Comment by tobous** [comment]
@srossbach Yes, I can create a pull request with the patch (probably tomorrow). But, for future reference, you should also have the rights to do it yourself if you would prefer. Our current practice is to push a new branch (based on the master) with the commit to the GitHub repository and then create a pull request from the branch onto the master.

**Comment by tobous** [comment]
I create a matching PR: #237

While testing the code, I noticed that I often (but not always) got an exception at the end of the session on both the host and the client side:
```
ERROR 21:24:52,858 [main] (ThreadUtils.java:40) Internal Error:
java.lang.NullPointerException
	at de.fu_berlin.inf.dpp.editor.EditorListener.unbind(EditorListener.java:108)
	at de.fu_berlin.inf.dpp.editor.EditorPool.remove(EditorPool.java:246)
	at de.fu_berlin.inf.dpp.editor.EditorPool.removeAllEditors(EditorPool.java:309)
	at de.fu_berlin.inf.dpp.editor.EditorManager.uninitialize(EditorManager.java:1826)
	at de.fu_berlin.inf.dpp.editor.EditorManager.access$17(EditorManager.java:1800)
	at de.fu_berlin.inf.dpp.editor.EditorManager$5$2.run(EditorManager.java:374)
	at de.fu_berlin.inf.dpp.util.ThreadUtils$1.run(ThreadUtils.java:38)
	at org.eclipse.swt.widgets.RunnableLock.run(RunnableLock.java:35)
	at org.eclipse.swt.widgets.Synchronizer.runAsyncMessages(Synchronizer.java:135)
	at org.eclipse.swt.widgets.Display.runAsyncMessages(Display.java:3563)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:3212)
	at org.eclipse.ui.internal.Workbench.runEventLoop(Workbench.java:2701)
	at org.eclipse.ui.internal.Workbench.runUI(Workbench.java:2665)
	at org.eclipse.ui.internal.Workbench.access$4(Workbench.java:2499)
	at org.eclipse.ui.internal.Workbench$7.run(Workbench.java:679)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:332)
	at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:668)
	at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:149)
	at org.eclipse.ui.internal.ide.application.IDEApplication.start(IDEApplication.java:123)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:196)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:110)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:79)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:344)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:179)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:622)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:577)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1410)
	at org.eclipse.equinox.launcher.Main.main(Main.java:1386)
ERROR 21:24:52,858 [main] (ThreadUtils.java:41) Original caller:
StackTrace:
	at de.fu_berlin.inf.dpp.util.ThreadUtils.wrapSafe(ThreadUtils.java:32)
	at de.fu_berlin.inf.dpp.ui.util.SWTUtils.runSafeSWTSync(SWTUtils.java:188)
	at de.fu_berlin.inf.dpp.editor.EditorManager$5.sessionEnded(EditorManager.java:371)
	at de.fu_berlin.inf.dpp.session.SarosSessionManager.sessionEnded(SarosSessionManager.java:723)
	at de.fu_berlin.inf.dpp.session.SarosSessionManager.stopSession(SarosSessionManager.java:350)
	at de.fu_berlin.inf.dpp.ui.util.CollaborationUtils$2.run(CollaborationUtils.java:158)
	at de.fu_berlin.inf.dpp.util.ThreadUtils$1.run(ThreadUtils.java:38)
	at java.lang.Thread.run(Thread.java:745)
```
I'm not sure whether this is connected to the same issue.

In general, there still seem to be a lot of problems with the partial sharing in Saros/E.
I now also got a new issue when trying to create a new file ("Created.java") in a directory that should have been shared (and whose parent directory should have been shared as well):
```
ERROR 21:10:48,133 [main] (SarosSession.java:829) PSFIC - unable to update partial sharing state, parent is not shared for file: L/alp/src/de/fu_berlin/alp4/Created.java (EclipseFileImpl)
```

**Comment by srossbach** [comment]
Yes I had seen this stacktrace as well. Cannot remember such a stacktrace from the past when leaving or stopping the session.

As for your issue. It seems that you only shared one file and not the folder that the file belongs to. Before I added my checks it was possible to add a file, however it is currently no longer possible.

We could either remove the check or start a finally a discussion with a final result on how partial sharing should work.


FYI I added this check to prevent this issue:


Alice shares file A.

Bob has also file B in the same place. Now Alice creates file B. Without this check it would be silently overriden on Bobs side.

**Comment by tobous** [comment]
Regarding the mentioned issue:
I re-tested this and made sure, that the parent directory is shared. Setup was as follows:
```
alp
 |-> src
      |-> de
           |-> fu_berlin
                |-> alp4
                     |-> new_base
                          |-> new_non_shared
                          |-> new_shared
```
All packages below `fu_berlin` contained at least one java file. `new_non_shared_base` (and its content) is the only directory that is marked as non-shared in the "Share project(s)..." dialog.
I then tried to create a file ("Testing.java") in the package `alp4` as the host and got the following error again:
```
ERROR 20:19:43,490 [main] (SarosSession.java:829) PSFIC - unable to update partial sharing state, parent is not shared for file: L/alp/src/de/fu_berlin/alp4/Testing.java (EclipseFileImpl)
```
Here is a bigger section of the log:
```
TRACE 20:19:43,481 [ModalContext] (SharedResourcesManager.java:234) received resource delta 'ResourceDelta(/)' contains:
C(0) /
C(0) /alp
C(0) /alp/src
C(0) /alp/src/de
C(0) /alp/src/de/fu_berlin
C(0) /alp/src/de/fu_berlin/alp4
A /alp/src/de/fu_berlin/alp4/Testing.java

TRACE 20:19:43,483 [ModalContext] (SharedResourcesManager.java:282) generated resource activities for current resource delta 'ResourceDelta(/)' : [FileActivity [dst:path=SPath [project=P/alp (EclipseProjectImpl), projectRelativePath=src/de/fu_berlin/alp4/Testing.java], src:path=N/A, type=CREATED, encoding=N/A, content=54 byte(s)]]
ERROR 20:19:43,490 [main] (SarosSession.java:829) PSFIC - unable to update partial sharing state, parent is not shared for file: L/alp/src/de/fu_berlin/alp4/Testing.java (EclipseFileImpl)
```

I also thought about the issue of creating resources that are already present for other participants. I noticed it while thinking about the whole Saros/I sub-modules not being shared thing and the resulting problems when deleting/moving/renaming directories that contain sub-modules on a different participant's side. But failing silently (for the user) seems weird to me. A resolution on the other participant's side would be better. But that is just my own opinion.

I think a general discussion on how partial sharing should work is a good idea, especially since it is going to be the default (and only) mode after Switches from sharing projects to sharing reference points (directories). This is currently the master's thesis of Oliver Hanßen. See issue #177. I am not sure in what state his work currently is.

**Comment by srossbach** [comment]
The initial partial sharing logic is found in an UI widget.

https://github.com/saros-project/saros/blob/master/de.fu_berlin.inf.dpp/src/de/fu_berlin/inf/dpp/ui/widgets/viewer/project/BaseResourceSelectionComposite.java#L427

https://github.com/saros-project/saros/blob/master/de.fu_berlin.inf.dpp/src/de/fu_berlin/inf/dpp/ui/widgets/viewer/project/BaseResourceSelectionComposite.java#L574

TLDR:

It is not that easy to share folders in partial sharing mode. You must not select a single file or you are out of luck.

**Comment by tobous** [comment]
I'm sorry, but I don't think I can deal with even more broken stuff right now. I am still working on the Saros/I filesystem stuff and fixing the other two Core problems (#223, #233).

Does this new issue influence the PR #237? Otherwise, if you don't have any objections, I would merge that PR and create a new GitHub issue for the creation stuff (or for partial sharing in general if you think there are also more/other problems; feel free to create such an issue yourself if you feel like it :wink:).

**Comment by tobous** [comment]
Fixed by #237
